File: backend/schemas/message.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class ConversationParticipantCreate(BaseModel):
    user_id: int
    participant_type: str = "user"
    display_name: Optional[str] = None
    role: Optional[str] = None


class ConversationParticipantResponse(BaseModel):
    id: int
    user_id: Optional[int] = None
    participant_type: str = "user"
    display_name: str
    role: Optional[str] = None
    last_read_at: Optional[datetime] = None
    joined_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ConversationCreate(BaseModel):
    name: str
    category: str = "direct"
    participants: list[ConversationParticipantCreate] = Field(default_factory=list)


class ConversationResponse(BaseModel):
    id: int
    name: str
    category: str
    last_message: Optional[str] = None
    last_message_at: Optional[datetime] = None
    unread_count: int
    created_at: Optional[datetime] = None
    participant_count: int = 0
    participants: list[ConversationParticipantResponse] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class MessageCreate(BaseModel):
    conversation_id: int
    sender_name: Optional[str] = None
    sender_role: Optional[str] = None
    sender_user_id: Optional[int] = None
    sender_participant_type: Optional[str] = None
    content: str
    message_type: str = "text"
    is_self: bool = False
    client_message_id: Optional[str] = None  # for deduplication on retry


class MessageResponse(BaseModel):
    id: int
    conversation_id: int
    sender_name: str
    sender_role: Optional[str] = None
    sender_user_id: Optional[int] = None
    sender_participant_type: Optional[str] = None
    content: str
    message_type: str
    is_self: bool
    is_deleted: bool = False
    edited_at: Optional[datetime] = None
    client_message_id: Optional[str] = None  # echo back for dedup
    created_at: datetime

    model_config = {"from_attributes": True}


# ── NEW: Message read receipt response ──────────────────────────
class MessageReadResponse(BaseModel):
    id: int
    message_id: int
    conversation_id: int
    user_id: int
    participant_type: str
    display_name: str
    read_at: datetime

    model_config = {"from_attributes": True}


# ── NEW: Notification preference schemas ────────────────────────
class NotificationPreferenceCreate(BaseModel):
    muted: bool = False
    mute_until: Optional[datetime] = None
    mention_only: bool = False
    push_enabled: bool = True


class NotificationPreferenceResponse(BaseModel):
    id: int
    user_id: int
    participant_type: str
    conversation_id: Optional[int] = None
    muted: bool
    mute_until: Optional[datetime] = None
    mention_only: bool
    push_enabled: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── NEW: Block list schemas ──────────────────────────────────────
class BlockUserRequest(BaseModel):
    user_id: int


class BlockedUserResponse(BaseModel):
    user_id: int
    display_name: Optional[str] = None
    blocked_at: Optional[datetime] = None

File: frontend_client/src/types/message.ts
export type ConversationCategory = "team" | "resident" | "alerts" | "direct" | string;

export type ConversationParticipant = {
  id: number;
  userId?: number;
  participantType: string;
  displayName: string;
  role?: string;
  lastReadAt?: string | null;
  joinedAt?: string | null;
};

export type ConversationItem = {
  id: string;
  contactId: string;
  name: string;
  role: string;
  category: ConversationCategory;
  preview: string;
  time: string;
  lastMessageAt?: string;
  unread: number;
  online: boolean;
  initials: string;
  avatarColor: string;
  participantCount?: number;
  participants: ConversationParticipant[];
  otherParticipant?: ConversationParticipant | null;
  targetUserId?: number;
};

export type ChatMessage = {
  id: string;
  conversationId: string;
  sender: "me" | "other";
  text: string;
  time: string;
  name?: string;
  senderRole?: string;
  senderUserId?: number;
  createdAt?: string;
  clientMessageId?: string;
  pending?: boolean;
};

export type BackendConversationParticipant = {
  id: number;
  user_id?: number | null;
  participant_type: string;
  display_name: string;
  role?: string | null;
  last_read_at?: string | null;
  joined_at?: string | null;
};

export type BackendConversation = {
  id: number;
  name: string;
  category: ConversationCategory;
  last_message?: string | null;
  last_message_at?: string | null;
  unread_count?: number;
  participant_count?: number;
  participants?: BackendConversationParticipant[];
};

export type BackendMessage = {
  id: number;
  conversation_id: number;
  sender_name: string;
  sender_role?: string | null;
  sender_user_id?: number | null;
  sender_participant_type?: string | null;
  content: string;
  message_type: string;
  is_self: boolean;
  created_at: string;
  client_message_id?: string | null;
};

export type CreateConversationInput = {
  name: string;
  category?: ConversationCategory;
  participants: Array<{
    userId: number;
    participantType?: string;
    displayName?: string;
    role?: string;
  }>;
};